### PR TITLE
backup: add s3 backend

### PR DESCRIPTION
**Pull request by xiang90**
/cc @hongchaodeng

### CODE REVIEW

File: pkg/backup/s3.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		return fmt.Errorf("failed to save snapshot: %v", err)
+	}
+	rc.Close()
+
+	// S3 put is atomic, so let's go ahead and put the key directly.
+	err = sb.S3.Put(key, tmpfile)
+	if err != nil {
+		return err
+	}
+	log.Printf("saved backup %s (size: %d) successfully", key, n)
```
Comment (hongchaodeng) [inline pkg/backup/s3.go]: use logrus?

Comment (xiang90) [inline pkg/backup/s3.go]: will do

File: pkg/backup/s3.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+	defer func() {
+		tmpfile.Close()
+		os.Remove(tmpfile.Name())
+	}()
+
+	n, err := io.Copy(tmpfile, rc)
+	if err != nil {
+		return fmt.Errorf("failed to save snapshot: %v", err)
+	}
+	rc.Close()
```
Comment (hongchaodeng) [inline pkg/backup/s3.go]: This will double close().

Comment (xiang90) [inline pkg/backup/s3.go]: will io.copy close it for us?

Comment (xiang90) [inline pkg/backup/s3.go]: io copy takes reader not closer. so why this is double close?

Comment (hongchaodeng) [inline pkg/backup/s3.go]: This is closed outside in writeSnap()

Comment (xiang90) [inline pkg/backup/s3.go]: ok. probably we should pass in just reader then. i will clean this up after this pr.

**Comment by xiang90** [comment]
@hongchaodeng All fixed. PTAL.

**Comment by hongchaodeng** [comment]
lgtm

**Pull request merged**